drivers/can/can_sam.c

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (c) 2021 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/can.h>
+#include <soc.h>
+#include <kernel.h>
+#include "can_sam.h"
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(can_driver, CONFIG_CAN_LOG_LEVEL);
+#define DT_DRV_COMPAT atmel_sam_can
+
+int can_sam_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+
+	*rate = SOC_ATMEL_SAM_MCK_FREQ_HZ / (CONFIG_CAN_SAM_CKDIV + 1);
+	return 0;
+}
+
+void can_sam_set_state_change_callback(const struct device *dev,
+				       can_state_change_callback_t isr)
+{
+	struct can_sam_data *data = DEV_DATA(dev);
+
+	data->mcan_data.state_change_cb = isr;
+}
+
+void can_sam_clock_enable(const struct can_sam_config *cfg)
+{
+	REG_PMC_PCK5 = PMC_PCK_CSS_PLLA_CLK | PMC_PCK_PRES(CONFIG_CAN_SAM_CKDIV);
+	PMC->PMC_SCER |= PMC_SCER_PCK5;
+	soc_pmc_peripheral_enable(cfg->pmc_id);
+}
+
+int can_sam_init(const struct device *dev)
+{
+	const struct can_sam_config *cfg = DEV_CFG(dev);
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+	struct can_sam_data *data = DEV_DATA(dev);
```

Comment:
Please, do not use `DEV_DATA()`/`DEV_CFG()` in new code. See #37616 for details.